src/layouts/BaseLayout/hooks/useUserAccessGate.js

```diff
@@ -31,47 +34,88 @@ function useOnboardingRedirect({ username }) {
+    isFetching: userIsFetching,
+  } = useUser({
     suspense: false,
-    enabled: !!provider || !config.IS_SELF_HOSTED,
+    enabled: !!provider && !config.IS_SELF_HOSTED,
   })
 
-  useOnboardingRedirect({ username: data?.user.username })
+  useOnboardingRedirect({ username: userData?.user?.username })
 
-  const isGuest = !data && isSuccess
-  let showAgreeToTerms = false,
-    showDefaultOrgSelector = false
+  const {
+    data: internalUser,
+    isLoading: internalUserIsLoading,
+    isSuccess: internalUserIsSuccess,
+    isFetching: internalUserIsFetching,
+  } = useInternalUser({
+    suspense: false,
+    enabled: !config.IS_SELF_HOSTED,
+  })
+
+  const foundUser = !userData && userIsSuccess
+  const foundInternalUser = !internalUser && internalUserIsSuccess
+  const isGuest = foundUser || foundInternalUser
+
+  let showAgreeToTerms = false
+  let showDefaultOrgSelector = false
+  let redirectToSyncPage = false
+
+  // the undefined provider check can be removed when the ToS has
+  // been refactored to no longer use a provider
+  if (
+    termsOfServicePage &&
+    !isUndefined(provider) &&
+    !isGuest &&
+    !config.IS_SELF_HOSTED
+  ) {
+    showAgreeToTerms = userData?.termsAgreement === false
+  }
 
-  // If we don't have a provider, we cant check the TOS agreement, so we alow the user to access the full experience.
-  if (!provider) {
-    return {
-      isFullExperience: true,
-      isLoading: false,
-      showAgreeToTerms,
-      showDefaultOrgSelector,
-    }
+  const onSyncPage = currentRoute.path === '/sync'
+  if (sentryLoginProvider && !isGuest && !onSyncPage) {
+    // owners array contains a list of the synced providers
+    // if it is zero then they haven't synced any other providers
+    redirectToSyncPage = isEqual(internalUser?.owners?.length, 0)
   }
 
-  if (termsOfServicePage && !isGuest) {
-    showAgreeToTerms = data?.termsAgreement === false
+  if (defaultOrgSelectorPage && !isUndefined(provider) && !isGuest) {
+    showDefaultOrgSelector = !userData?.owner?.defaultOrgUsername
   }
 
-  if (defaultOrgSelectorPage && !isGuest) {
-    showDefaultOrgSelector = !data?.owner?.defaultOrgUsername
+  // so when a query is disabled it goes into it's loading state which will be
+  // true on the /sync route, and well we don't really care about that call
+  // so this just ignores that fact and only checks to see if the internal user
+  // is loading rather then both ... since we won't be able to fetch both.
+  let isLoading = internalUserIsLoading && internalUserIsFetching
+  if (!isUndefined(provider)) {
+    isLoading =
+      (userIsLoading && userIsFetching) ||
+      (internalUserIsLoading && internalUserIsFetching)
   }
 
   // Not fully tested logic yet, waiting on API to be available.
   // Assuming self hosted users do not need to sign
   return {
     isFullExperience:
-      !!config.IS_SELF_HOSTED || (!showAgreeToTerms && !showDefaultOrgSelector),
+      !!config.IS_SELF_HOSTED ||
```

Comment:
you're already checking if it's self hosted here no need to check for self hosted in App.jsx

Reply:
So we may want to enable some of these features inside `useUserAccessGate` to self hosted users potentially, so if that was removed and also removed in `App.jsx` it could potentially show the page accidentally. I'd prefer to keep it the way it is so it will explicitly not show while running in self-hosted mode.

Reply:
would we ever show sync page for self hosted though?

Reply:
Yes, I believe the eventual plan is to allow other OAuth providers on self-hosted